sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -219,29 +242,57 @@ def incr_mean_variance_axis0(X, last_mean, last_var, unsigned long last_n):
     `utils.extmath._batch_mean_variance_update`.
 
     """
-    cdef unsigned long n_samples = X.shape[0]
-    cdef unsigned int n_features = X.shape[1]
+    if X.dtype != np.float32:
+        X = X.astype(np.float64)
+    return _incr_mean_variance_axis0(X.data, X.shape, X.indices, X.indptr,
+                                     last_mean, last_var, last_n)
+
+
+def _incr_mean_variance_axis0(np.ndarray[floating, ndim=1] X_data,
```

Comment:
This should be a `cdef`  right? There is not point in exposing this as part of the Python-level API.

I think we also need the usual decorators:

```python
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
```

Maybe we could enable them globally for this cython file instead.